### PR TITLE
8334167: Test java/lang/instrument/NativeMethodPrefixApp.java timed out

### PR DESCRIPTION
**Pull request by jaikiran**
Can I please get a review of this test-only change which proposes to fix the test timeout reported in https://bugs.openjdk.org/browse/JDK-8334167?

The JBS issue has comments which explains what causes the timeout. The commit in this PR addresses those issues by updating the test specific `ClassFileTransformer` to only instrument application specific class instead of all (core) classes. The test was introduced several years back to verify the feature introduced in https://bugs.openjdk.org/browse/JDK-6263319. As such, the proposed changes in this PR continue to test that feature - we now merely use application specific class' native method to verify the semantics of that feature.

Additional cleanups have been done in the test to make sure that if any exception does occur in the ClassFileTransformer then it does get recorded and that then causes the test to fail.

With this change, I have run tier1 through tier6 and the test passes. Additionally, without this change I've run the test with a test repeat of 100 with virtual threads enabled and the test hangs occasionally (as expected). With this proposed fix, I have then run the test with virtual threads, around 300 times and it hasn't failed or hung in any of those instances.

---------
### Progress
- [x] Change must be properly reviewed (1 review required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer))
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue

### Issue
 * [JDK-8334167](https://bugs.openjdk.org/browse/JDK-8334167): Test java/lang/instrument/NativeMethodPrefixApp.java timed out (**Bug** - P4)

### Reviewers
 * [David Holmes](https://openjdk.org/census#dholmes) (@dholmes-ora - **Reviewer**)
 * [Alan Bateman](https://openjdk.org/census#alanb) (@AlanBateman - **Reviewer**) 🔄 Re-review required (review applies to [1d303c90](https://git.openjdk.org/jdk/pull/20154/files/1d303c903aa6ff9896aab2adb5c86dc0ff637ec9))
 * [Serguei Spitsyn](https://openjdk.org/census#sspitsyn) (@sspitsyn - **Reviewer**)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jdk.git pull/20154/head:pull/20154` \
`$ git checkout pull/20154`

Update a local copy of the PR: \
`$ git checkout pull/20154` \
`$ git pull https://git.openjdk.org/jdk.git pull/20154/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 20154`

View PR using the GUI difftool: \
`$ git pr show -t 20154`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jdk/pull/20154.diff">https://git.openjdk.org/jdk/pull/20154.diff</a>

</details>

### Webrev
[Link to Webrev Comment](https://git.openjdk.org/jdk/pull/20154#issuecomment-2225186316)

### CODE REVIEW

**Review by dholmes-ora — COMMENTED** [review]
@jaikiran there is a lot of unrelated refactoring and style changes here that obscures what the actual fix is.

File: test/jdk/java/lang/instrument/NativeMethodPrefixApp.java [hunk 1]
```diff
@@ -34,27 +31,26 @@
...
- *          java.instrument
+ * @modules java.instrument
  * @library /test/lib
  * @build bootreporter.StringIdCallback bootreporter.StringIdCallbackReporter
  *        asmlib.Instrumentor NativeMethodPrefixAgent
  * @enablePreview
  * @comment The test uses asmlib/Instrumentor.java which relies on ClassFile API PreviewFeature.
- * @run driver/timeout=240 NativeMethodPrefixApp roleDriver
+ * @run main/native/timeout=240 NativeMethodPrefixApp roleDriver
  * @comment The test uses a higher timeout to prevent test timeouts noted in JDK-6528548
```
Comment (AlanBateman) [inline test/jdk/java/lang/instrument/NativeMethodPrefixApp.java:44]: Is /timeout=240 (and the comment) needed now. If I read the old issue correctly it was due to use a JDK mounted on a network file system.

**Review by AlanBateman — COMMENTED** [review]
I think the changes looks okay. It narrows to instrumenting a specific test class, thus removing all the side effects of the transformer code generating wrapper methods for JDK classes. The cleanups looks okay,

File: test/jdk/java/lang/instrument/NativeMethodPrefixApp.java (hunk 1 above)
Comment (jaikiran) [reply test/jdk/java/lang/instrument/NativeMethodPrefixApp.java]: In https://github.com/openjdk/jdk/pull/19495#discussion_r1625470816 we had considered removing the higher timeout from this test. But then decided to let it stay at that time. I think we don't need it anymore, not even when the test runs with `-Xcomp`. The recent runs in our CI for this test haven't shown the necessity of using this higher timeout. I have updated the PR to remove this. I plan to run additional tests in our CI to be sure that the removal of the timeout doesn't cause unexpected failures.

**Comment by jaikiran** [comment]
Hello David,

> @jaikiran there is a lot of unrelated refactoring and style changes here that obscures what the actual fix is.

I've updated the PR to undo some of the cosmetic changes that were introduced to cleanup the code a bit. The updated state of the PR now just includes functional updates to the test to address the issue noted in the linked JBS.

**Comment by jaikiran** [comment]
> Would it be possible to at least re-format the declaration of the "transform" method as it's very messy and hard to see what is declaration vs. body. 

Done.

**Review by AlanBateman — APPROVED** [review]
I don't have any other comments, it's good to have this test focus on just wrapping one native method.

File: test/jdk/java/lang/instrument/NativeMethodPrefixApp.java [hunk 2]
```diff
@@ -34,27 +31,25 @@
...
+ * @run main/native NativeMethodPrefixApp roleDriver
  */
 public class NativeMethodPrefixApp implements StringIdCallback {
 
     // This test is fragile like a golden file test.
     // It assumes that a specific non-native library method will call a specific
     // native method.  The below may need to be updated based on library changes.
-    static String goldenNativeMethodName = "getStartupTime";
-
+    static String goldenNativeMethodName = "fooBarNativeMethod";
```
Comment (dholmes-ora) [inline test/jdk/java/lang/instrument/NativeMethodPrefixApp.java:50]: The comment block above seems no longer applicable now this is not a library class method.

**Review by dholmes-ora — APPROVED** [review]
Okay. Still seems to be a lot more changes than needed for the key update, but so be it.

Thanks

File: test/jdk/java/lang/instrument/NativeMethodPrefixApp.java (hunk 2 above)
Comment (jaikiran) [reply test/jdk/java/lang/instrument/NativeMethodPrefixApp.java:49]: I've now updated the code comment to note what that constant is for.

**Review by dholmes-ora — APPROVED** [review]

**Review by sspitsyn — APPROVED** [review]
Nice test update making it more reliable and readable.
Looks good.

**Comment by jaikiran** [comment]
Thank you all for the reviews and inputs. I'll go ahead with the integration in a few hours from now, once the CI test jobs I trigger for this, complete.

**Pull request closed without merging**